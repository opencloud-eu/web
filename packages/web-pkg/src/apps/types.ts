import { App, ComponentCustomProperties, Ref } from 'vue'
import { RouteLocationRaw, Router, RouteRecordRaw } from 'vue-router'
import { Extension, ExtensionPoint } from '../composables/piniaStores'
import { IconFillType } from '../helpers'
import { Resource, SpaceResource } from '@opencloud-eu/web-client'
import { Translations } from 'vue3-gettext'

export interface AppReadyHookArgs {
  globalProperties: ComponentCustomProperties & Record<string, any>
  router: Router
  instance?: App
  portal?: any
}

export interface AppNavigationItem {
  isActive?: () => boolean
  activeFor?: { name?: string; path?: string }[]
  isVisible?: () => boolean
  fillType?: IconFillType
  icon?: string
  name: string | (() => string)
  route?: RouteLocationRaw
  handler?: () => void
  priority?: number
}

export type AppConfigObject = Record<string, any>

export interface ApplicationFileExtension {
  app?: string
  extension?: string
  createFileHandler?: (arg: {
    fileName: string
    space: SpaceResource
    currentFolder: Resource
  }) => Promise<Resource>
  hasPriority?: boolean
  label?: string | (() => string)
  name?: string
  icon?: string
  mimeType?: string
  newFileMenu?: { menuTitle: () => string }
  routeName?: string
  secureView?: boolean
}

/** ApplicationInformation describes required information of an application */
export interface ApplicationInformation {
  color?: string
  id?: string
  name?: string
  icon?: string
  iconFillType?: IconFillType
  iconColor?: string
  img?: string
  meta?: {
    fileSizeLimit?: number
  }
  extensions?: ApplicationFileExtension[]
  defaultExtension?: string
  translations?: Translations
}

/**
 * ApplicationTranslations is a map of language keys to translations
 */
export interface ApplicationTranslations {
  [lang: string]: {
    [key: string]: string | string[]
  }
}

/** ClassicApplicationScript reflects classic application script structure */
export interface ClassicApplicationScript {
  appInfo?: ApplicationInformation
  routes?: ((args: ComponentCustomProperties) => RouteRecordRaw[]) | RouteRecordRaw[]
  navItems?: ((args: ComponentCustomProperties) => AppNavigationItem[]) | AppNavigationItem[]
  translations?: ApplicationTranslations
  extensions?: Ref<Extension[]>
  extensionPoints?: Ref<ExtensionPoint<any>[]>
  initialize?: () => void
  ready?: (args: AppReadyHookArgs) => Promise<void> | void
  mounted?: (args: AppReadyHookArgs) => void
  // TODO: move this to its own type
  setup?: (args: { applicationConfig: AppConfigObject }) => ClassicApplicationScript
}

export type ApplicationSetupOptions = { applicationConfig: AppConfigObject }

export const defineWebApplication = (args: {
  setup: (options: ApplicationSetupOptions) => ClassicApplicationScript
}) => {
  return args
}
