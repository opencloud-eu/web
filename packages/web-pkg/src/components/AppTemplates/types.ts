import { Resource } from '@opencloud-eu/web-client'
import { AppConfigObject } from '../../apps/types'
import { Ref } from 'vue'

export interface AppWrapperSlotArgs {
  applicationConfig: AppConfigObject
  resource: Resource
  currentContent: Ref<string>
  isDirty: boolean
  isReadOnly: boolean
  url: string
}
