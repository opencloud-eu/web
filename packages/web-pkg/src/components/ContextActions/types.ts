import { Action } from '../../composables'

export type MenuSectionDrop = {
  label: string
  name: string
  icon: string
  items?: Action[]
  emptyMessage?: string
}

export type MenuSection = {
  name: string
  items?: Action[]
  dropItems?: MenuSectionDrop[]
}
