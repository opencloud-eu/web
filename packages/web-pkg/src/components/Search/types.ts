import { SearchResource } from '@opencloud-eu/web-client'
import { Component } from 'vue'

export interface GenericSearchResultItem {
  id: string
  name: string
  icon?: string
}

export interface SearchResultValue {
  id: string
  // FIXME: SearchResource is very specific for webdav search
  data: GenericSearchResultItem | SearchResource
}

export interface SearchResult {
  totalResults: number | null
  values: SearchResultValue[]
}

export interface SearchList {
  component: Component // Component props need to match SearchResultValue data type

  search(term: string): Promise<SearchResult>
}

export interface SearchPreview extends SearchList {
  available: boolean
}

export interface SearchProvider {
  id: string
  available: boolean
  displayName?: string
  previewSearch?: SearchPreview
  listSearch?: SearchList
}
