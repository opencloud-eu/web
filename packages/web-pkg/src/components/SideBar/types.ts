import { defineComponent } from 'vue'
import { IconFillType } from '../../helpers/resource'
import { Item } from '@opencloud-eu/web-client'

export interface SideBarPanelContext<R extends Item, P extends Item, T extends Item> {
  root?: R
  parent?: P
  items?: T[]
}

export interface SideBarPanel<R extends Item, P extends Item, T extends Item> {
  name: string
  icon: string
  iconFillType?: IconFillType
  title(context: SideBarPanelContext<R, P, T>): string
  isVisible(context: SideBarPanelContext<R, P, T>): boolean
  component: ReturnType<typeof defineComponent>
  componentAttrs?(context: SideBarPanelContext<R, P, T>): any
  /**
   * defines if the panel is a `root` level panel in the right sidebar.
   * In the long run this should be configured by admins or even users, as it's ideally not to be decided by extension developer.
   */
  isRoot?(context: SideBarPanelContext<R, P, T>): boolean
}
