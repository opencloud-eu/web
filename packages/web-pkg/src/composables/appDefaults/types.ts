import { MaybeRef } from '../../utils'
import { LocationQuery } from '../router'
import { SpaceResource } from '@opencloud-eu/web-client'
import { RouteParams } from 'vue-router'

export interface FileContext {
  path: MaybeRef<string>
  driveAliasAndItem: MaybeRef<string>
  space: MaybeRef<SpaceResource>
  item: MaybeRef<string>
  itemId: MaybeRef<string>
  fileName: MaybeRef<string>
  routeName: MaybeRef<string>
  routeParams: MaybeRef<RouteParams>
  routeQuery: MaybeRef<LocationQuery>
}
