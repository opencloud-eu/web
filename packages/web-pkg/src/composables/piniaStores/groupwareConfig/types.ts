import { z } from 'zod'

export const MailCapabilitiesSchema = z.object({
  maxMailboxDepth: z.number(),
  maxSizeMailboxName: z.number(),
  maxMailboxesPerEmail: z.number(),
  maxSizeAttachmentsPerEmail: z.number(),
  mayCreateTopLevelMailbox: z.boolean(),
  maxDelayedSend: z.number()
})

export const SieveCapabilitiesSchema = z.object({
  maxSizeScriptName: z.number(),
  maxSizeScript: z.number(),
  maxNumberScripts: z.number(),
  maxNumberRedirects: z.number()
})

export const AccountCapabilitiesSchema = z.object({
  mail: MailCapabilitiesSchema.optional(),
  sieve: SieveCapabilitiesSchema.optional()
})

export const IdentitySchema = z.object({
  id: z.string(),
  name: z.string(),
  email: z.string().email(),
  mayDelete: z.boolean()
})

export const AccountSchema = z.object({
  name: z.string(),
  isPersonal: z.boolean(),
  isReadOnly: z.boolean(),
  capabilities: AccountCapabilitiesSchema,
  identities: z.array(IdentitySchema)
})

export const LimitsSchema = z.object({
  maxSizeUpload: z.number(),
  maxConcurrentUpload: z.number(),
  maxSizeRequest: z.number(),
  maxConcurrentRequests: z.number()
})

export const AccountsSchema = z.record(z.string(), AccountSchema)

export const PrimaryAccountsSchema = z.record(z.string(), z.string())

export const RawGroupwareConfigSchema = z.object({
  version: z.string(),
  capabilities: z.array(z.string()),
  limits: LimitsSchema,
  accounts: AccountsSchema,
  primaryAccounts: PrimaryAccountsSchema
})

export type RawGroupwareConfig = z.infer<typeof RawGroupwareConfigSchema>
