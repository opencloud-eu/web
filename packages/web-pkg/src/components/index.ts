export * from './AppBar'
export * from './AppTemplates'
export * from './ContextActions'
export * from './FilesList'
export * from './Filters'
export * from './Modals'
export * from './SideBar'
export * from './Search'
export * from './Spaces'
export * from './TextEditor'

export { default as AppLoadingSpinner } from './AppLoadingSpinner.vue'
export { default as AppTopBar } from './AppTopBar.vue'
export { default as BatchActions } from './BatchActions.vue'
export { default as CustomComponentTarget } from './CustomComponentTarget.vue'
export { default as ItemFilter } from './ItemFilter.vue'
export { default as ItemFilterToggle } from './ItemFilterToggle.vue'
export { default as LinkRoleDropdown } from './LinkRoleDropdown.vue'
export { default as LoadingIndicator } from './LoadingIndicator.vue'
export { default as NoContentMessage } from './NoContentMessage.vue'
export { default as Pagination } from './Pagination.vue'
export { default as QuotaSelect } from './QuotaSelect.vue'
export { default as SpaceQuota } from './SpaceQuota.vue'
export { default as SearchBarFilter } from './SearchBarFilter.vue'
export { default as ViewOptions } from './ViewOptions.vue'
export { default as PortalTarget } from './PortalTarget.vue'
export { default as CreateShortcutModal } from './CreateShortcutModal.vue'
export { default as CreateLinkModal } from './CreateLinkModal.vue'
