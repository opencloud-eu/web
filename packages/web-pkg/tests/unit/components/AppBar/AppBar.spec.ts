import { ref } from 'vue'
import AppBar from '../../../../src/components/AppBar/AppBar.vue'
import { mock } from 'vitest-mock-extended'
import { Resource, SpaceResource } from '@opencloud-eu/web-client'
import {
  defaultComponentMocks,
  defaultPlugins,
  shallowMount,
  RouteLocation,
  PartialComponentProps
} from '@opencloud-eu/web-test-helpers'
import { ArchiverService } from '../../../../src/services'
import { FolderView } from '../../../../src/ui/types'
import { useExtensionRegistry, ViewOptions } from '../../../../src'
import { OcBreadcrumb } from '@opencloud-eu/design-system/components'

const selectors = {
  ocBreadcrumbStub: 'oc-breadcrumb-stub',
  batchActionsStub: 'batch-actions-stub',
  viewOptionsStub: 'view-options-stub',
  sidebarToggleStub: 'sidebar-toggle-stub',
  mobileNavPortal: 'portal-target[name="app.runtime.mobile.nav"]'
}

const selectedFiles = [mock<Resource>(), mock<Resource>()]
const actionSlot = "<button class='action-slot'>Click</button>"
const contentSlot = "<div class='content-slot'>Foo</div>"

const breadcrumbItems = [
  { text: 'Example1', to: '/' },
  { text: 'Example2', to: '/foo' }
]
const breadCrumbItemWithContextActionAllowed = {
  text: 'Example Special',
  to: '/bar',
  allowContextActions: true
}

describe('AppBar component', () => {
  describe('renders', () => {
    it('by default no breadcrumbs, no bulkactions, no sharesnavigation but viewoptions and sidebartoggle', () => {
      const { wrapper } = getShallowWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
    describe('breadcrumbs', () => {
      it('if given, by default without breadcrumbsContextActionsItems', () => {
        const { wrapper } = getShallowWrapper([], {}, { breadcrumbs: breadcrumbItems })
        expect(wrapper.find(selectors.ocBreadcrumbStub).exists()).toBeTruthy()
        expect(
          wrapper.findComponent<typeof OcBreadcrumb>(selectors.ocBreadcrumbStub).props('items')
        ).toEqual(breadcrumbItems)
      })
      it('if given, with breadcrumbsContextActionsItems if allowed on last breadcrumb item', () => {
        const { wrapper } = getShallowWrapper(
          [],
          {},
          { breadcrumbs: [...breadcrumbItems, breadCrumbItemWithContextActionAllowed] }
        )
        expect(wrapper.find(selectors.ocBreadcrumbStub).exists()).toBeTruthy()
        expect(
          wrapper.findComponent<typeof OcBreadcrumb>(selectors.ocBreadcrumbStub).props('items')
        ).toEqual([...breadcrumbItems, breadCrumbItemWithContextActionAllowed])
      })
      it('not if no breadcrumb items given', () => {
        const { wrapper } = getShallowWrapper([], {}, { breadcrumbs: [] })
        expect(wrapper.find(selectors.ocBreadcrumbStub).exists()).toBeFalsy()
      })
      it('not if one breadcrumb item is given in mobile view', () => {
        const { wrapper } = getShallowWrapper(
          [],
          {},
          { breadcrumbs: [breadcrumbItems[0]] },
          mock<RouteLocation>({ name: '' }),
          true
        )
        expect(wrapper.find(selectors.ocBreadcrumbStub).exists()).toBeFalsy()
      })
    })
    describe('bulkActions', () => {
      it('if enabled', () => {
        const { wrapper } = getShallowWrapper(selectedFiles, {}, { hasBulkActions: true })
        expect(wrapper.find(selectors.batchActionsStub).exists()).toBeTruthy()
      })
      it('if 1 file selected on trash routes', () => {
        const { wrapper } = getShallowWrapper(
          [selectedFiles[0]],
          {},
          { hasBulkActions: true },
          mock<RouteLocation>({
            name: 'files-trash-generic',
            path: '/files/trash/personal/admin'
          })
        )
        expect(wrapper.find(selectors.batchActionsStub).exists()).toBeTruthy()
      })
    })
    describe('mobile navigation portal', () => {
      it.each([
        { items: [], shows: true },
        { items: [breadcrumbItems[0]], shows: true },
        { items: [breadcrumbItems[0], breadcrumbItems[1]], shows: false }
      ])('if less than 2 breadcrumb items given', ({ items, shows }) => {
        const { wrapper } = getShallowWrapper([], {}, { breadcrumbs: items })
        expect(wrapper.find(selectors.mobileNavPortal).exists()).toBe(shows)
      })
    })
    describe('viewoptions', () => {
      it('show if options are available', () => {
        const { wrapper } = getShallowWrapper([], {}, { hasViewOptions: true })
        expect(wrapper.find(selectors.viewOptionsStub).exists()).toBeTruthy()
      })
      it('hide if options are not available', () => {
        const { wrapper } = getShallowWrapper([], {}, { hasViewOptions: false })
        expect(wrapper.find(selectors.viewOptionsStub).exists()).toBeFalsy()
      })
      it('passes viewModes array to ViewOptions', () => {
        const viewModes = [mock<FolderView>()]
        const { wrapper } = getShallowWrapper([], {}, { hasViewOptions: true, viewModes })
        expect(
          wrapper.findComponent<typeof ViewOptions>(selectors.viewOptionsStub).props('viewModes')
        ).toEqual(viewModes)
      })
    })
    it('if given, with content in the actions slot', () => {
      const { wrapper } = getShallowWrapper([], { actions: actionSlot })
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('if given, with content in the content slot', () => {
      const { wrapper } = getShallowWrapper([], { content: contentSlot })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})

function getShallowWrapper(
  selected: Resource[] = [],
  slots = {},
  props: PartialComponentProps<typeof AppBar> = {
    breadcrumbs: [],
    viewModes: [],
    hasBulkActions: false,
    hasViewOptions: true
  },
  currentRoute = mock<RouteLocation>({
    name: 'files-spaces-generic',
    path: '/files/spaces/personal/admin'
  }),
  isMobileWidth = false
) {
  const plugins = defaultPlugins({
    piniaOptions: {
      resourcesStore: { resources: selected, selectedIds: selected.map(({ id }) => id) }
    }
  })

  const { requestExtensions } = useExtensionRegistry()
  vi.mocked(requestExtensions).mockReturnValue([])

  const mocks = {
    ...defaultComponentMocks({
      currentRoute
    }),
    $archiverService: mock<ArchiverService>()
  }
  mocks.$route.meta.title = 'ExampleTitle'

  return {
    wrapper: shallowMount(AppBar, {
      props: { ...props, space: mock<SpaceResource>() },
      slots,
      global: {
        plugins,
        provide: { ...mocks, isMobileWidth: ref(isMobileWidth) },
        mocks
      }
    })
  }
}
