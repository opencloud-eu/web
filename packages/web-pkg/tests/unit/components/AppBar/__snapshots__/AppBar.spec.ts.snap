// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppBar component > renders > by default no breadcrumbs, no bulkactions, no sharesnavigation but viewoptions and sidebartoggle 1`] = `
"<div data-v-aba8ddeb="" id="files-app-bar" class="px-4 bg-role-surface rounded-t-xl [display:inherit] top-0 sticky">
  <div data-v-aba8ddeb="" class="files-topbar py-2">
    <h1 data-v-aba8ddeb="" class="oc-invisible-sr">ExampleTitle</h1>
    <oc-hidden-announcer-stub data-v-aba8ddeb="" announcement="No items selected." level="polite"></oc-hidden-announcer-stub>
    <div data-v-aba8ddeb="" class="flex items-center files-app-bar-controls min-h-13 justify-end">
      <!--v-if-->
      <portal-target data-v-aba8ddeb="" name="app.runtime.mobile.nav"></portal-target>
      <!--v-if-->
      <div data-v-aba8ddeb="" id="files-app-bar-controls-right" class="flex">
        <view-options-stub data-v-aba8ddeb="" perpagestorageprefix="files" hashiddenfiles="true" hasfileextensions="true" haspagination="true" paginationoptions="20,50,100,250,500" perpagequeryname="items-per-page" perpagedefault="100" viewmodedefault="resource-table" viewmodes=""></view-options-stub>
      </div>
    </div>
    <div data-v-aba8ddeb="" class="files-app-bar-actions flex items-center justify-end mt-1 min-h-10 gap-2">
      <div data-v-aba8ddeb="" class="flex-1 flex justify-start items-center">
        <!--v-if-->
      </div>
    </div>
  </div>
</div>"
`;

exports[`AppBar component > renders > if given, with content in the actions slot 1`] = `
"<div data-v-aba8ddeb="" id="files-app-bar" class="px-4 bg-role-surface rounded-t-xl [display:inherit] top-0 sticky">
  <div data-v-aba8ddeb="" class="files-topbar py-2">
    <h1 data-v-aba8ddeb="" class="oc-invisible-sr">ExampleTitle</h1>
    <oc-hidden-announcer-stub data-v-aba8ddeb="" announcement="No items selected." level="polite"></oc-hidden-announcer-stub>
    <div data-v-aba8ddeb="" class="flex items-center files-app-bar-controls min-h-13 justify-end">
      <!--v-if-->
      <portal-target data-v-aba8ddeb="" name="app.runtime.mobile.nav"></portal-target>
      <!--v-if-->
      <div data-v-aba8ddeb="" id="files-app-bar-controls-right" class="flex">
        <view-options-stub data-v-aba8ddeb="" perpagestorageprefix="files" hashiddenfiles="true" hasfileextensions="true" haspagination="true" paginationoptions="20,50,100,250,500" perpagequeryname="items-per-page" perpagedefault="100" viewmodedefault="resource-table" viewmodes=""></view-options-stub>
      </div>
    </div>
    <div data-v-aba8ddeb="" class="files-app-bar-actions flex items-center justify-end mt-1 min-h-10 gap-2">
      <div data-v-aba8ddeb="" class="flex-1 flex justify-start items-center"><button class="action-slot">Click</button>
        <!--v-if-->
      </div>
    </div>
  </div>
</div>"
`;

exports[`AppBar component > renders > if given, with content in the content slot 1`] = `
"<div data-v-aba8ddeb="" id="files-app-bar" class="px-4 bg-role-surface rounded-t-xl [display:inherit] top-0 sticky">
  <div data-v-aba8ddeb="" class="files-topbar py-2">
    <h1 data-v-aba8ddeb="" class="oc-invisible-sr">ExampleTitle</h1>
    <oc-hidden-announcer-stub data-v-aba8ddeb="" announcement="No items selected." level="polite"></oc-hidden-announcer-stub>
    <div data-v-aba8ddeb="" class="flex items-center files-app-bar-controls min-h-13 justify-end">
      <!--v-if-->
      <portal-target data-v-aba8ddeb="" name="app.runtime.mobile.nav"></portal-target>
      <!--v-if-->
      <div data-v-aba8ddeb="" id="files-app-bar-controls-right" class="flex">
        <view-options-stub data-v-aba8ddeb="" perpagestorageprefix="files" hashiddenfiles="true" hasfileextensions="true" haspagination="true" paginationoptions="20,50,100,250,500" perpagequeryname="items-per-page" perpagedefault="100" viewmodedefault="resource-table" viewmodes=""></view-options-stub>
      </div>
    </div>
    <div data-v-aba8ddeb="" class="files-app-bar-actions flex items-center justify-end mt-1 min-h-10 gap-2">
      <div data-v-aba8ddeb="" class="flex-1 flex justify-start items-center">
        <!--v-if-->
      </div>
    </div>
    <div class="content-slot">Foo</div>
  </div>
</div>"
`;
